packages/x-data-grid-pro/src/hooks/features/serverSideData/serverSideInterfaces.ts

```diff
@@ -0,0 +1,63 @@
+import { GridRowId } from '@mui/x-data-grid';
+import { GridGetRowsParams, GridGetRowsResponse } from '../../../models';
+
+/**
+ * The dataSource API interface that is available in the grid [[apiRef]].
+ */
+export interface GridDataSourceApi {
+  /**
+   * Set the loading state of a parent row.
+   * @param {string} parentId The id of the parent node.
+   * @param {boolean} loading The loading state to set.
+   */
+  setChildrenLoading: (parentId: GridRowId, loading: boolean) => void;
+  /**
+   * Set error occured while fetching the children of a row.
+   * @param {string} parentId The id of the parent node.
+   * @param {Error} error The error of type `Error` or `null`.
+   */
+  setChildrenFetchError: (parentId: GridRowId, error: Error | null) => void;
+  /**
+   * Fetch/refetch the top level rows.
+   */
+  fetchTopLevelRows: () => void;
+  /**
+   * Enqueues the fetch of the children of a row.
+   * @param {GridRowId} id The id of the rowNode belonging to the group to be fetched.
+   */
+  enqueueChildrenFetch: (id: GridRowId) => void;
+}
+
+export interface GridDataSourcePrivateApi {
+  /**
+   * Initiates the fetch of the children of a row.
+   * @param {string} id The id of the rowNode belonging to the group to be fetched.
+   */
+  fetchRowChildren: (id: GridRowId) => void;
+  /**
+   * Resets the server side state.
+   */
+  resetServerSideState: () => void;
+}
+
+/**
+ * The server side cache API interface that is available in the grid [[apiRef]].
+ */
+export interface GridServerSideCacheApi {
```

Comment:
A thing I don't like about our API is that we put all the methods in the same namespace, so it can be confusing just from the name what a method refers to.

Reply:
Could you think of a better (possibly non-breaking) way? 🤔

Reply:
We can start grouping methods when adding new ones. What do you think?

Reply:
What would be the ideal grouping strategy, based on feature/hook name?
How does something like `apiRef.current.dataSource.fetchTopLevelRows()` look?

Reply:
Yes, I think grouping by feature would make sense and be easier to understand for users.

Btw can we make those also `unstable_dataSource` to match the props?

Reply:
On thinking more about this, it might be confusing for some users too.

For example, if the `apiRef.current.unstable_dataSourceCache` returns the cache object, it might be expected to get the data source object with `apiRef.current.unstable_dataSource`.
So we might think about doing something like **\`${featureName}Api\`** in the namespace name. Something like `dataSourceApi`, it will never be misunderstood I think.

Another thing to consider is, having same or different namespaces for private API, for example having namespace `dataSourceApiPrivate` for the private methods vs using the same namespace for all the methods.

We may also have to upgrade our docs infrastructure for [gridApi](https://deploy-preview-12317--material-ui-x.netlify.app/x/api/data-grid/grid-api/)https://deploy-preview-12317--material-ui-x.netlify.app/x/api/data-grid/grid-api/ to support grouped methods.

What do you think?

Reply:
> For example, if the apiRef.current.unstable_dataSourceCache returns the cache object, it might be expected to get the data source object with apiRef.current.unstable_dataSource.

We can put `cache` into `apiRef.current.unstable_dataSourceCache.cache`. What do you think?

> Another thing to consider is, having same or different namespaces for private API, for example having namespace dataSourceApiPrivate for the private methods vs using the same namespace for all the methods.

I don't have a strong opinion on this one.

Reply:
Or at `apiRef.current.unstable_dataSource.cache`.

Reply:
I moved cache logic to `useGridDataSource` and cache is accessible at `apiRef.current.unstable_dataSource.cache`

Reply:
I haven't yet moved the private methods to the namespace `unstable_dataSource` namespace, as I am unsure what the best strategy would be.
We could handle it later once we have a solid one since they are meant for internal use.
If you have any thoughts, feel free to share.

Reply:
I'd put them on the same namespace, but it doesn't matter much since we can refactor them anytime, so it's fine as it is.